api.js

```diff
@@ -0,0 +1,164 @@
+"use strict";
+
+const {classes: Cc, interfaces: Ci, results: Cr, utils: Cu} = Components;
+
+Cu.import("resource://gre/modules/XPCOMUtils.jsm");
+
+Cu.import("resource://devtools/shared/event-emitter.js");
+Cu.import("resource://gre/modules/ExtensionUtils.jsm");
+const {SingletonEventManager} = ExtensionUtils;
+
+XPCOMUtils.defineLazyModuleGetter(this, "webrtcUI",
+                                  "resource:///modules/webrtcUI.jsm");
+
+let origReceiveMessage = null;
+let boundRecv = null;
+let pcBlockers = new Set();
+let hookCount = 0;
+let emitter = new EventEmitter();
+
+function receiveMessage(msg) {
+
+  switch (msg.name) {
+  case "rtcpeer:Request": {
+    let origin = msg.target.contentPrincipal.origin;
+    let id = msg.data.callID;
+    emitter.emit("peer-request", origin, id);
+
+    function answer(reply) {
+      msg.target.messageManager.sendAsyncMessage(reply, {
+        callID: msg.data.callID,
+        windowID: msg.data.windowID,
+      });
+    }
+
+    let blockers = Array.from(pcBlockers);
+    function next() {
+      if (blockers.length == 0) {
+        origReceiveMessage(msg);
+        return;
+      }
+
+      let blocker = blockers.shift();
+      return blocker(origin, id).then(result => {
+        if (result == "allow") {
+          answer("rtcpeer:Allow");
+        } else if (result == "deny") {
+          answer("rtcpeer:Deny");
+        } else {
+          return next();
+        }
+      });
+    }
+    next().catch(err => {
+      Cu.reportError(`error in PeerConnection blocker: ${err.message}`);
+      origReceiveMessage(msg);
+    });
+    break;
+  }
+
+  case "rtcpeer:CancelRequest": {
+    let origin = msg.target.contentPrincipal.origin;
+    let id = msg.data;
+    emitter.emit("cancel-peer-request", origin, id);
+    origReceiveMessage(msg);
+    break;
+  }
+
+  case "webrtc:UpdateBrowserIndicators": {
+    // Beware, this will need to change when https://bugzil.la/1299577 lands
+    let origin = msg.target.contentPrincipal.origin;
+    let {camera, microphone} = msg.data;
+    emitter.emit("media-permissions", origin, {camera, microphone});
+  }
+
+  default:
+    origReceiveMessage(msg);
+  }
+}
+
+function hook() {
+  let first = (hookCount == 0);
+  hookCount++;
+  if (first) {
+    origReceiveMessage = webrtcUI.receiveMessage;
+    webrtcUI.receiveMessage = receiveMessage;
+  }
+}
```

Comment:
If hookCount > 0 then we previously hooked our version of receiveMessage in.  If webrtcUI.receiveMessage is not our receiveMessage then somebody else has also hooked receiveMessage.  I think the only reasonable thing is to assume that the other hook is a good citizen and is passing incoming messages through to us when appropriate.
This of course can/will get cleaner if we improve the hooking facilities in webrtcUI.jsm as discussed in a previous comment.